staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,623 @@
+// filter that there has been a timing splinter and the filter
+// re-calcuates the priority level to use); in all other cases
+// `quiescent` will be returned `false` (even if the system is
+// quiescent by then).  In the non-quiescent unhappy cases the
+// request is eventually rejected, which means to return with
+// `execute=false`.  In the happy case the caller is required to
+// invoke the returned `afterExecution` after the request is done
+// executing.  The hash value and hand size are used to do the
+// shuffle sharding.
+func (qs QueueSetImpl) Wait(hashValue uint64, handSize int32) (quiescent, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify what should/shouldn't be locked!!!!
+	// TODO(aaron-prindle) collapse all of FQ into one layer/lock (vs 3)
+	//   currently able to collapse to 1 impl layer and 2 locks...
+
+	qs.LockAndSyncTime()
+	// TODO(aaron-prindle) verify and test quiescent
+	// A call to Wait while the system is quiescent will be rebuffed by
+	// returning `quiescent=true`.
+	if qs.Quiescent {
+		return true, false, func() {}
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a packet and enqueue
+	pkt := qs.TimeoutOldRequestsAndRejectOrEnqueue(hashValue, handSize)
+	// pkt == nil means that the request was rejected - no remaining
+	// concurrency shares and at max queue length already
+	if pkt == nil {
+		return false, false, func() {}
+	}
+	// ========================================================================
+
+	// ------------------------------------------------------------------------
+	// Step 2:
+	// 1) The next step is to invoke the method that dequeues as much as possible.
+
+	// This method runs a loop, as long as there
+	// are non-empty queues and the number currently executing is less than the
+	// assured concurrency value.  The body of the loop uses the fair queuing
+	// technique to pick a queue, dequeue the request at the head of that
+	// queue, increment the count of the number executing, and send `{true,
+	// handleCompletion(that dequeued request)}` to the request's channel.
+	qs.DequeueWithChannelAsMuchAsPossible()
+	// ------------------------------------------------------------------------
+	qs.Lock.Unlock()
+
+	// ************************************************************************
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` on either request timeout or receipt of a record on the
+	// newly arrived request's channel, and return appropriately.  If a record
+	// has been sent to the request's channel then this `select` will
+	// immediately complete
+	channelPkt := pkt.(*Packet)
```

Comment:
To support testing with a FakeEventClock, the QueueSet will need to hold not only a clock but also a `*sync.WaitGroup`, on which `Done()` will need to be called here to indicate that this goroutine has done as much work as it can until something unblocks it.

Reply:
updated

Reply:
updated, including nil check.